presto-main/src/main/java/com/facebook/presto/server/ServerInfoResource.java

```diff
@@ -14,30 +14,82 @@
 package com.facebook.presto.server;
 
 import com.facebook.presto.client.ServerInfo;
+import com.facebook.presto.spi.NodeState;
 
 import javax.inject.Inject;
 import javax.ws.rs.GET;
+import javax.ws.rs.POST;
 import javax.ws.rs.Path;
 import javax.ws.rs.Produces;
+import javax.ws.rs.WebApplicationException;
 import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.Response;
 
+import static com.facebook.presto.server.ResourceUtil.assertRequest;
+import static com.facebook.presto.spi.NodeState.ACTIVE;
+import static com.facebook.presto.spi.NodeState.SHUTTING_DOWN;
+import static java.lang.String.format;
 import static java.util.Objects.requireNonNull;
+import static javax.ws.rs.core.MediaType.APPLICATION_JSON;
+import static javax.ws.rs.core.MediaType.TEXT_PLAIN;
+import static javax.ws.rs.core.Response.Status.BAD_REQUEST;
 
 @Path("/v1/info")
 public class ServerInfoResource
 {
     private final ServerInfo serverInfo;
+    private final GracefulShutdownHandler shutdownHandler;
 
     @Inject
-    public ServerInfoResource(ServerInfo serverInfo)
+    public ServerInfoResource(ServerInfo serverInfo, GracefulShutdownHandler shutdownHandler)
     {
         this.serverInfo = requireNonNull(serverInfo, "serverInfo is null");
+        this.shutdownHandler = requireNonNull(shutdownHandler, "shutdownHandler is null");
     }
 
     @GET
-    @Produces(MediaType.APPLICATION_JSON)
+    @Produces(APPLICATION_JSON)
     public ServerInfo getServerInfo()
     {
         return serverInfo;
     }
+
+    @POST
+    @Path("state")
+    @Produces(TEXT_PLAIN)
+    public Response updateState(NodeState state)
+            throws WebApplicationException
+    {
+        switch (state) {
+            case SHUTTING_DOWN:
+                assertRequest(!shutdownHandler.isShutdownRequested(), "Shutdown in progress ...");
+                shutdownHandler.requestShutdown();
+                return Response.ok().build();
+            case ACTIVE:
+            case INACTIVE:
+                throw new WebApplicationException(Response
+                        .status(BAD_REQUEST)
+                        .type(MediaType.TEXT_PLAIN)
+                        .entity(format("Invalid state transition to %s", state))
+                        .build());
+            default:
+                return Response.status(BAD_REQUEST)
+                        .type(TEXT_PLAIN)
+                        .entity(format("Invalid state %s", state))
+                        .build();
+        }
+    }
+
+    @GET
+    @Path("state")
+    @Produces(APPLICATION_JSON)
```

Comment:
Oh, I see why you have this as json. The `RemoteNodeState` takes JSON. I think that's fine